### ISSUE TITLE
allow many_single_char_names is broken

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust-clippy#3198]
[Playground](https://play.rust-lang.org/?gist=28a08c4f14d03b96bd3e61954b879b1d&version=stable&mode=debug&edition=2015):

```rust
#[cfg_attr(
    feature = "cargo-clippy",
    allow(clippy::many_single_char_names)
)]
pub fn append_to() {
    let (a, b, c, d, e) = (0, 0, 0, 0, 0);
}
```

produces

```
warning: 5th binding whose name is just one char
 --> src/main.rs:7:22
  |
7 |     let (a, b, c, d, e) = (0, 0, 0, 0, 0);
  |                      ^
  |
  = note: #[warn(clippy::many_single_char_names)] on by default
  = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#many_single_char_names
```

That lint should not trigger, since it is allowed for the function.

**Comment by flip1995** [comment]
Applying the allow to a function does not change the behavior of the lint inside the function. This is weird, since the lint is produced through `EarlyLintPass::check_[impl_]item()` and a function is an item.

It works if you turn the `cfg_attr` into an inner attribute. Probably something you don't want to do.

Removing the `clippy::` and checking this with `cargo +stable clippy` doesn't produce a warning. I'm afraid that this could have something to do with `tool_lints`...

**Comment by gnzlbg** [comment]
I'm hitting this issue with other lints like `shadow_unrelated` as well. Moving the cfg_attr into an inner attribute does not work around this form me, can you show on the play ground how you are doing that ? This still produces the warning for me:

```rust
pub fn append_to() {
    #![cfg_attr(
        feature = "cargo-clippy",
        allow(clippy::many_single_char_names)
    )]
    let (a, b, c, d, e) = (0, 0, 0, 0, 0);
}
```

**Comment by flip1995** [comment]
Sure: [Playground](https://play.rust-lang.org/?gist=8d925f61305fc39d6d2536cef0a0ca0a&version=nightly&mode=debug&edition=2015)
```rust
#![feature(tool_lints)]

#![cfg_attr(
    feature = "cargo-clippy",
    allow(clippy::many_single_char_names)
)]

#[allow(unused_variables)] 
pub fn main() {
    let (a, b, c, d, e) = (0, 0, 0, 0, 0);
}
```

**Comment by flip1995** [comment]
Hm maybe something changed how lint attributes are processed... Since when do you have this issues?

**Comment by gnzlbg** [comment]
Ah, you meant making it an inner attribute of the scope where the function is, yeah I'd prefer not to do that. I've just disabled the lint globally as a workaround for now.

> Since when do you have these issues?

I just started updating criterion to the latest clippy, and the last Rust nightly version criterion used was from june... so... I'd guess between june and today, which isn't probably very helpful.

**Comment by flip1995** [comment]
I'm looking into that!

**Comment by flip1995** [comment]
This error started with `nightly-2018-08-24`. The previous `nightly-2018-08-19` does not produce a warning for your example. (That means, it is not connected to the `tool_lints` :tada:)

**Comment by flip1995** [comment]
One of the commits are at fault here:

https://github.com/rust-lang/rust/commit/758239c9c9e00b7c0daba6496ef246a85e7b5ce5
https://github.com/rust-lang/rust/commit/d2048b6db375299b681d4f4728b8e7cad9f74d5f
https://github.com/rust-lang/rust/commit/6bf6d50a6ff7685b4aa09172d9d09f03f250da9d

Travis: https://travis-ci.org/flip1995/test_allow_on_fn

**Comment by flip1995** [comment]
This is no longer an issue: [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a86c78a6080bd08af29292eb88c2cfd8)

**Comment by gnzlbg** [comment]
Is there a test that prevents this from breaking again?

**Comment by flip1995** [comment]
Oh good point!
